### ISSUE TITLE
Handle embedded (INLINE) resources as attachments if the cid identifier doesn't occur in the HTML body

### ISSUE THREAD

**Issue by jkxyx205** [bbottema/simple-java-mail#179]
Hi, Benny Bottema

I have a problem, the details are as follows。

**1. Scene description**

  Here is my [hello.eml](https://github.com/bbottema/simple-java-mail/files/2523957/hello.eml.zip) file which include a attachment and a embedded image exported by client application [Foxmail](https://www.foxmail.com/win/en/). Open with outlook screenshot below:

<img width="399" alt="a" src="https://user-images.githubusercontent.com/5242715/47630433-b3df0d00-db7b-11e8-839f-1d5e4c0606c1.png">

**2. Parse eml but can't get any information**
```java
Email oldEmail = EmailConverter.emlToEmail("/Users/rick/jkxyx205/tmp/hello.eml");
```
<img width="580" alt="333" src="https://user-images.githubusercontent.com/5242715/47630786-9d39b580-db7d-11e8-95f1-eef3b1e524d7.png">

**Comment by bbottema** [comment]
Can you perform the same check for EmailConverter.emlToMimeMessage() please?

**Comment by jkxyx205** [comment]
I can't parse any thing either.

```java
return new MimeMessage(session, new ByteArrayInputStream(eml.getBytes(UTF_8)));
```
Maybe `java.mail` does not have the ability to parse this file but outlook can.

**Comment by bbottema** [comment]
Hmm, maybe this EML requires a specific javax.mail version? I'll give it a try as well, but I'm affraid this is a limitation of the underlying Java Mail framework.

**Comment by bbottema** [comment]
I'm getting a completely empty result without errors.

For what it's worth, the EML is not validating in any of the following:

- [tools.gbif.org/dwca-validator/eml](https://tools.gbif.org/dwca-validator/eml.do)
- [knb.ecoinformatics.org/emlparser](https://knb.ecoinformatics.org/emlparser)
- [tools.ietf.org/tools/msglint](https://tools.ietf.org/tools/msglint/)

But [mimevalidator.net](https://www.mimevalidator.net/index.html) thinks it's fine.

**Comment by bbottema** [comment]
Sorry, I got confused. It's working fine for me.

Are you sure your variable `eml` refers to the _content_ of the file and not the file's _path_.

Here's my result:

```
InputStream resourceAsStream = EmailHelper.class.getClassLoader().getResourceAsStream("test-messages/hello.eml");
Email e =  EmailConverter.emlToEmail(resourceAsStream);
```
![image](https://user-images.githubusercontent.com/2115718/47656694-d5e28a80-db8f-11e8-8989-a727783e688a.png)

**Comment by jkxyx205** [comment]
Yes, you are right, what a ridiculous mistake I did.

But I found another problem, lost a embeded image as I metioned [#173](https://github.com/bbottema/simple-java-mail/issues/173), but 2 attachments

![image](https://user-images.githubusercontent.com/5242715/47657798-b6b61e00-dbcc-11e8-8b50-ede31a916e58.png)

**Comment by bbottema** [comment]
If an embedded attachment isn't actually embedded it is treated as an attachment. Is it being used in the body?

**Comment by jkxyx205** [comment]
Outlook treat the mail as a attachment and a embedded images.

Ok,I will continue to track it later. Thank you for your help, simple-java-mail excellent framework!!!

**Comment by jkxyx205** [comment]
Hmm, finally I got the root cause.

When export EML by client foxmail  which does't  specify header `Content-Disposition: inline; filename=image.png`  when it is a embedded image. If the `disposition` is not provided, the part be treated as attachment by **simple-java-mail**. 

See **org.simplejavamail.converter.internal.mimemessage.MimeMessageParser.java**
```java
private static void parseMimePartTree(@Nonnull final MimePart currentPart, @Nonnull final ParsedMimeMessageComponents parsedComponents) {
		for (final Header header : retrieveAllHeaders(currentPart)) {
			parseHeader(header, parsedComponents);
		}
		
		final String disposition = parseDisposition(currentPart);
		
		if (isMimeType(currentPart, "text/plain") && parsedComponents.plainContent == null && !Part.ATTACHMENT.equalsIgnoreCase(disposition)) {
			parsedComponents.plainContent = parseContent(currentPart);
		} else if (isMimeType(currentPart, "text/html") && parsedComponents.htmlContent == null && !Part.ATTACHMENT.equalsIgnoreCase(disposition)) {
			parsedComponents.htmlContent = parseContent(currentPart);
		} else if (isMimeType(currentPart, "multipart/*")) {
			final Multipart mp = parseContent(currentPart);
			for (int i = 0, count = countBodyParts(mp); i < count; i++) {
				parseMimePartTree(getBodyPartAtIndex(mp, i), parsedComponents);
			}
		} else {
			final DataSource ds = createDataSource(currentPart);
			// If the diposition is not provided, the part should be treated as attachment
			if (disposition == null || Part.ATTACHMENT.equalsIgnoreCase(disposition)) {
				parsedComponents.attachmentList.put(parseResourceName(parseContentID(currentPart), parseFileName(currentPart)), ds);
			} else if (Part.INLINE.equalsIgnoreCase(disposition)) {
				if (parseContentID(currentPart) != null) {
					parsedComponents.cidMap.put(parseContentID(currentPart), ds);
				} else {
					// contentID missing -> treat as standard attachment
					parsedComponents.attachmentList.put(parseResourceName(null, parseFileName(currentPart)), ds);
				}
			} else {
				throw new IllegalStateException("invalid attachment type");
			}
		}
	}
```

If I add `Content-Disposition: inline; filename=image.png` manually, it works fine. ˆ-ˆ

<img width="1063" alt="aaaaa" src="https://user-images.githubusercontent.com/5242715/47693868-3debac00-dc36-11e8-8a37-e96c5464b26d.png">


So my question is coming, why  can't I treated as embedded image if `disposition` is not provided，behaves normally like Outlook.

**Comment by jkxyx205** [comment]
BTW, Outlook, Mac Mail, Foxmail can parse hello.eml correctly, one attachment + one embedded image.

**Comment by bbottema** [comment]
> So my question is coming, why can't I treated as embedded image if disposition is not provided，behaves normally like Outlook.

The spec says the following about default Content-Disposition in case of absence:
```
The Content-Disposition Header Field

   Content-Disposition is an optional header field. In its absence, the
   MUA may use whatever presentation method it deems suitable.
```
So spec-wise, we're free to do as we see fit. However, I'm unsure of the best handling here. So all the clients you mention don't parse it 'correctly' so much as rather how they see fit. It's all correct.

Here's what I'm going to do: treat all resources with missing Content-Disposition header as INLINE. Then if an INLINE resource (like an embedded image) does _not_ occur in the HTML body (ie. cid:myImage), treat it as an attachment instead.

This also treats embedded images with proper INLINE disposition as attachment if the image is not actually used in the HTML body.

**Comment by bbottema** [comment]
@jkxyx205, I've released a new SNAPSHOT with the fix. Can you please verify (you'll need to [add the snapshot repo](https://stackoverflow.com/a/7717234/441662)).

**Comment by jkxyx205** [comment]
I tested it, and the performance was as good as I expected. It was really great.

**Comment by bbottema** [comment]
Released in 5.1.0
